<a href="https://colab.research.google.com/github/JacoboHH27/Bussines-Analitics/blob/main/1_Arbol_de_desicion_clase.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Cargamos las librerias


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#se trae el modelo
from sklearn.tree import DecisionTreeClassifier

#Para llamar los archivos del drive sin subirlos - Sin tener que subirlos
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


1. Se cargan los datos desde el drive

In [ ]:
nxl= '/content/drive/MyDrive/Analitica de Negocios/0. SolicitantesCrédito(USD) (1).xlsx'
XDB= pd.read_excel(nxl, sheet_name=1)
XDB=XDB.dropna()
XDB.head()

#seleccionamos las variables de interes
XD=XDB[['Edad', 'Ingresos', 'Egresos', 'Monto (EAD)']]
XD.head()
XD=np.array(XD) #Convierte las variables en numeros - variables de entrada
yd=XDB[['PreApr']]; yd=np.array(yd) #variable de salida  - convierto a la tabla


2.Se procede con la construccion del modelo de arbol

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)
ydp=mar.predict(XD) #estos valores sirven para evaluar el proyecto

#evaluamos el modelo
from sklearn.metrics import confusion_matrix
confusion_matrix(yd,ydp)
cm=confusion_matrix(yd,ydp)
VN=cm[0,0]; FN=cm[1,0]; VP=cm[1,1]; FP=cm[0,1]

Ex=(VP+VN)/(VP+VN+FP+FN)
Se=VP/(VP+FN); Es=VN/(VN+FP)

print('Exactitud=',Ex)
print('Sensibilidad=',Se)
print('Especificidad=',Es)

Exactitud= 0.7771311194796303
Sensibilidad= 0.7766215747485259
Especificidad= 0.7776275768840825


3.Se procede con la visualización del arbol

In [ ]:
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data #que datos le voy a colocar al arbol

vs=['Edad','Ingresos','Egresos','Monto (EAD)'] #estos nombres aparecen en el arbol

dot_data=export_graphviz(mar, feature_names=vs)
graph=graph_from_dot_data(dot_data)
graph.write_png('arbol.png') #entre mas individuos tenga el nodo y el gini sea menor, sera un mejor Nodo

True

4. Le calculamos los valores de pre aprobación a los solicitantes nuevos

In [ ]:
XDB2=pd.read_excel(nxl, sheet_name=2)
XDB2= XDB2[['Edad', 'Ingresos', 'Egresos', 'Monto (EAD)']]
XDB2=XDB2.dropna()

XD2=np.array(XDB2)
ydp2=mar.predict(XDB2)
XDB2['PreApr']=ydp2
XDB2.head()

XDB2.to_excel('NuevosSolicitantesPreApr.xlsx')

np.sum(ydp2)
len(ydp2)-np.sum(ydp2)

1276

**analizis de resultado final**
De acuerdo con la base de datos se puede evidenciar un total de 5842 solicitantes de creditos de los cuales 2959 eran pre aprobados que representan un 51% y un total de 2883 prenegados 49%. De acuerdo con lo anterior el modelo de arbol arrojo un modelo de exactitud cercano al 77%, así como un indice de sensibilidad cercano al 77% igualmente y que muestran el comportamineto del modelo frente al inidice de espicificidad el cual mide el comportamiento del modelo en la clasificación de solicitantes PreNegados.
Es importante mencionar que un modelo que exhiba valores de exactitud por encima del 75% es considerado un modelo aceptable para el pronostico de los datos.
Para la base de datos de solicitantes nuevos (2536 solicitantes) se puede observar que el modelo clasifico un total de 1260 preaprobados que corresponde al x% y 1276 en la categoria de prenegados. Lo anterior muestra la predominancia del modelo por la clasificación de la categoria de pre negados.